src/lit-element.ts

```diff
@@ -31,6 +33,112 @@ export class LitElement extends UpdatingElement {
+      const styles = this.styles;
+      // As a performance optimization to avoid duplicated styling that can
+      // occur especially when composing via subclassing, de-duplicate styles
+      // preserving the last item in the list. The last item is kept to
+      // try to preserve cascade order with the assumption that it's most
+      // important that last added styles override previous styles.
+      const styleSet = styles.reduceRight((set, s) => {
+        set.add(s);
+        return set;
+      }, new Set());
+      this._styles = Array.from(styleSet).reverse();
+    }
+    return this._styles;
+  }
+
+  /**
+   * Node or ShadowRoot into which element DOM should be rendered. Defaults
+   * to an open shadowRoot.
+   */
+  protected renderRoot?: Element|DocumentFragment;
+
+  /**
+   * Performs element initialization. By default this calls `createRenderRoot`
+   * to create the element `renderRoot` node and captures any pre-set values for
+   * registered properties.
+   */
+  protected initialize() {
+    super.initialize();
+    this.renderRoot = this.createRenderRoot();
+    // Note, if renderRoot is not a shadowRoot, styles would/could apply to the
+    // element's getRootNode(). While this could be done, we're choosing not to
+    // support this now since it would require different logic around de-duping.
+    if (window.ShadowRoot && this.renderRoot instanceof window.ShadowRoot) {
+      this.adoptStyles();
+    }
+  }
+
+  /**
+   * Returns the node into which the element should render and by default
+   * creates and returns an open shadowRoot. Implement to customize where the
+   * element's DOM is rendered. For example, to render into the element's
+   * childNodes, return `this`.
+   * @returns {Element|DocumentFragment} Returns a node into which to render.
+   */
+  protected createRenderRoot(): Element|ShadowRoot {
+    return this.attachShadow({mode : 'open'});
+  }
+
+  /**
+   * Applies styling to the element shadowRoot using the `static get styles`
+   * property. Styling will apply using `shadowRoot.adoptedStyleSheets` where
+   * available and will fallback otherwise. When Shadow DOM is polyfilled,
+   * ShadyCSS scopes styles and adds them to the document. When Shadow DOM
+   * is available but `adoptedStyleSheets` is not, styles are appended to the
+   * end of the `shadowRoot` to [mimic spec behavior](https://wicg.github.io/construct-stylesheets/#using-constructed-stylesheets).
+   */
+  protected adoptStyles() {
+    const styles = (this.constructor as typeof LitElement)._uniqueStyles;
+    if (styles.length === 0) {
+      return;
+    }
+    // There are three separate cases here based on Shadow DOM support.
+    // (1) shadowRoot polyfilled: use ShadyCSS
+    // (2) shadowRoot.adoptedStyleSheets available: use it.
+    // (3) shadowRoot.adoptedStyleSheets polyfilled: append styles after rendering.
+    if (window.ShadyCSS !== undefined && !window.ShadyCSS.nativeShadow) {
```

Comment:
Is it worth moving more of the prep steps to the static one-time spot (`_uniqueStyles`)?  Basically, move the `map` calls out of instance-time work, and possibly also create a template of styles to clone for the final case?